File: frontend/src/types/conversation.ts
/**
 * TypeScript types for conversation state and API responses.
 * Mirrors backend Pydantic models for type safety.
 */

// =============================================================================
// Message Types
// =============================================================================

/**
 * Message sender role.
 */
export type MessageRole = "user" | "assistant";

/**
 * Message in a conversation.
 */
export interface Message {
  id: number;
  role: MessageRole;
  content: string;
  created_at: string;
}

// =============================================================================
// Conversation Types
// =============================================================================

/**
 * Conversation summary for list view.
 */
export interface ConversationSummary {
  id: string;
  message_count: number;
  created_at: string;
  updated_at: string;
}

/**
 * Full conversation with messages.
 */
export interface ConversationDetail {
  id: string;
  messages: Message[];
  created_at: string;
  updated_at: string;
}

/**
 * Paginated conversation list response.
 */
export interface ConversationListResponse {
  conversations: ConversationSummary[];
  total: number;
  limit: number;
  offset: number;
}

// =============================================================================
// Error Types
// =============================================================================

/**
 * Conversation error codes.
 */
export type ConversationErrorCode =
  | "CONVERSATION_NOT_FOUND"
  | "MESSAGE_TOO_LONG"
  | "DATABASE_ERROR";

/**
 * Conversation error response.
 */
export interface ConversationErrorResponse {
  error_code: ConversationErrorCode;
  message: string;
  details?: Record<string, unknown>;
}

// =============================================================================
// Chat State Types (Frontend-specific)
// =============================================================================

/**
 * Global chat state managed by ChatContext.
 */
export interface ChatState {
  /** List of conversation summaries for sidebar */
  conversations: ConversationSummary[];

  /** Currently selected conversation ID */
  currentConversationId: string | null;

  /** Messages in current conversation */
  messages: Message[];

  /** Whether data is being loaded */
  isLoading: boolean;

  /** Whether a message is being sent/processed */
  isSending: boolean;

  /** Error message to display */
  error: string | null;
}

/**
 * Chat state actions.
 */
export type ChatAction =
  | { type: "SET_CONVERSATIONS"; payload: ConversationSummary[] }
  | { type: "ADD_CONVERSATION"; payload: ConversationSummary }
  | { type: "REMOVE_CONVERSATION"; payload: string }
  | { type: "UPDATE_CONVERSATION"; payload: ConversationSummary }
  | { type: "SELECT_CONVERSATION"; payload: string | null }
  | { type: "SET_MESSAGES"; payload: Message[] }
  | { type: "ADD_MESSAGE"; payload: Message }
  | { type: "SET_LOADING"; payload: boolean }
  | { type: "SET_SENDING"; payload: boolean }
  | { type: "SET_ERROR"; payload: string | null }
  | { type: "CLEAR_ERROR" }
  | { type: "NEW_CONVERSATION" };

/**
 * Initial chat state.
 */
export const initialChatState: ChatState = {
  conversations: [],
  currentConversationId: null,
  messages: [],
  isLoading: false,
  isSending: false,
  error: null,
};

File: backend/src/models/conversation.py
"""Pydantic models for conversation history storage."""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class MessageRole(str, Enum):
    """Message sender role."""

    USER = "user"
    ASSISTANT = "assistant"


class MessageCreate(BaseModel):
    """Internal model for creating messages (from orchestrator)."""

    conversation_id: str
    role: MessageRole
    content: str = Field(..., min_length=1, max_length=10000)


class MessageResponse(BaseModel):
    """Message in API response."""

    id: int
    role: MessageRole
    content: str
    created_at: datetime


class ConversationSummary(BaseModel):
    """Conversation in list response."""

    id: str
    message_count: int
    created_at: datetime
    updated_at: datetime


class ConversationDetail(BaseModel):
    """Full conversation with messages."""

    id: str
    messages: list[MessageResponse]
    created_at: datetime
    updated_at: datetime


class ConversationListResponse(BaseModel):
    """Paginated conversation list."""

    conversations: list[ConversationSummary]
    total: int
    limit: int
    offset: int


class ConversationErrorCode(str, Enum):
    """Error codes for conversation operations."""

    CONVERSATION_NOT_FOUND = "CONVERSATION_NOT_FOUND"
    MESSAGE_TOO_LONG = "MESSAGE_TOO_LONG"
    DATABASE_ERROR = "DATABASE_ERROR"


class ConversationErrorResponse(BaseModel):
    """Error response for conversation operations."""

    error_code: ConversationErrorCode
    message: str
    details: dict[str, Any] | None = None
